misc/scripts/prepare-db-upgrade.sh

```diff
@@ -0,0 +1,168 @@
+#!/bin/sh
+#
+# Prepare the upgrade and downgrade script directories for a DB Schema upgrade.
+
+set -e
+set -u
+
+app_name="$(basename "$0")"
+
+usage()
+{
+  exit_code="$1"
+  shift
+
+  cat >&2 <<EOF
+${app_name}: $@
+${app_name}: Generate skeleton upgrade script.
+Usage: ${app_name} [--prev_hash <COMMITISH>] [--lang <LANG>]"
+
+--prev-hash <COMMITISH>
+        Hash/branch to use to get SHA1 for previous DB scheme.
+        Default: origin/main
+
+--lang <LANG>
+        Language to update the schema for.
+
+Must be run within the git repo needing an update.
+EOF
+  exit "${exit_code}"
+}
+
+prev_hash="origin/main"
+lang="cpp"
+
+while [ $# -gt 0 ]; do
+  case "$1" in
+    -x)
+      set -x
+      ;;
+    -h | --help)
+      usage 0
+      ;;
+    --prev-hash)
+      if [ $# -eq 1 ]; then
+        usage 2 "--prev-hash requires Commit/Branch option"
+      fi
+      shift
+      prev_hash="$1"
+      ;;
+    --lang)
+      if [ $# -eq 1 ]; then
+        usage 2 "--lang requires a language option"
+      fi
+      shift
+      lang="$1"
+      ;;
+    --)
+      shift
+      break
+      ;;
+    -*)
+      usage 2 "Unrecognised option: $1"
+      ;;
+    *)
+      break
+      ;;
+  esac
+  shift
+done
+
+if [ $# -gt 0 ]; then
+  usage 2 "Unrecognised operand: $1"
+fi
+
+top_level="$(git rev-parse --show-superproject-working-tree)"
+
+if [ "x${top_level}" = "x" ]; then
+  top_level="$(git rev-parse --show-toplevel)"
+fi
+
+if [ "x${top_level}" = "x" ]; then
+  usage 2 "Not in a code repository."
+fi
+
+case "${lang}" in
+  java)
+    scheme_file="${lang}/ql/lib/config/semmlecode.dbscheme"
+    ;;
+  csharp | cpp | javascript | python)
+    scheme_file="${lang}/ql/lib/semmlecode.${lang}.dbscheme"
+    ;;
+  ruby)
+    scheme_file="${lang}/ql/lib/${lang}.dbscheme"
+    ;;
+  *)
+    usage 2 "Unrecognised language: ${lang}"
+    ;;
+esac
+
+qldir="${top_level}/ql"
```

Comment:
This should make it possible to run from either repository:
```suggestion
qldir="${top_level}"                                                                        
if [ ! -f "${qldir}/${scheme_file}" ]; then
  qldir="${top_level}/ql"
  if [ ! -f "${qldir}/${scheme_file}" ]; then                                               
    echo "Cannot find ${qldir}/${scheme_file}"                                             
    exit 2                                                                                  
  fi
fi
```

Reply:
Is the aim to support both `codeql` being checked out as a submodule and independently?

Reply:
That's right. I imagine someone working on C/C++ would likely run this from the submodule, while someone working on Ruby (where everything, including the extractor, is self-contained in the `codeql` repo) would run it from an independent checkout of `codeql`.

Reply:
I wonder if we should go for something simpler based on the script location:
```
app_dir="$(dirname "$0")"
cd ${app_dir}
qldir="$(git rev-parse --show-toplevel)"

if [ "x${qldir}" = "x" ]; then
  usage 2 "Script not in a code repository."
fi
```

Reply:
Yeah, that's certainly better.

Reply:
Done.